crates/matrix-sdk-crypto-js/src/machine.rs

```diff
@@ -0,0 +1,549 @@
+    /// responses need to be passed back to the state machine using
+    /// `mark_request_as_sent`.
+    #[wasm_bindgen(js_name = "outgoingRequests")]
+    pub fn outgoing_requests(&self) -> Promise {
+        let me = self.inner.clone();
+
+        future_to_promise(async move {
+            Ok(me
+                .outgoing_requests()
+                .await?
+                .into_iter()
+                .map(OutgoingRequest)
+                .map(TryFrom::try_from)
+                .collect::<Result<Vec<JsValue>, _>>()?
+                .into_iter()
+                .collect::<Array>())
+        })
+    }
+
+    /// Mark the request with the given request ID as sent (see
+    /// `outgoing_requests`).
+    ///
+    /// `request_id` represents the unique ID of the request that was
+    /// sent out. This is needed to couple the response with the now
+    /// sent out request. `response_type` represents the type of the
+    /// request that was sent out. `response` represents the response
+    /// that was received from the server after the outgoing request
+    /// was sent out. `
+    #[wasm_bindgen(js_name = "markRequestAsSent")]
+    pub fn mark_request_as_sent(
+        &self,
+        request_id: &str,
+        request_type: requests::RequestType,
+        response: &str,
+    ) -> Result<Promise, JsError> {
+        let transaction_id = OwnedTransactionId::from(request_id);
+        let response = response_from_string(response)?;
+        let incoming_response = responses::OwnedResponse::try_from((request_type, response))?;
+
+        let me = self.inner.clone();
+
+        Ok(future_to_promise(async move {
+            Ok(me.mark_request_as_sent(&transaction_id, &incoming_response).await.map(|_| true)?)
+        }))
+    }
+
+    /// Encrypt a room message for the given room.
+    ///
+    /// Beware that a group session needs to be shared before this
+    /// method can be called using the `share_group_session` method.
+    ///
+    /// Since group sessions can expire or become invalid if the room
+    /// membership changes, client authors should check with the
+    /// `should_share_group_session` method if a new group session
+    /// needs to be shared.
+    ///
+    /// `room_id` is the ID of the room for which the message should
+    /// be encrypted. `event_type` is the type of the event. `content`
+    /// is the plaintext content of the message that should be
+    /// encrypted.
+    ///
+    /// # Panics
+    ///
+    /// Panics if a group session for the given room wasn't shared beforehand.
+    #[wasm_bindgen(js_name = "encryptRoomEvent")]
+    pub fn encrypt_room_event(
+        &self,
+        room_id: &identifiers::RoomId,
+        event_type: &str,
+        content: &str,
+    ) -> Result<Promise, JsError> {
+        let room_id = room_id.inner.clone();
+        let content: Box<RawJsonValue> = serde_json::from_str(content)?;
+        let content = AnyMessageLikeEventContent::from_parts(event_type, &content)?;
```

Comment:
This doesn't support custom event fields nor custom events, `OlmMachine::encrypt_room_event_raw()` exists for that reason.

Reply:
Let's create a new issue ticket listing this issue item, as part of the next work to do to finish Wasm+JS.

Reply:
But this one is trivial, you deserialize into a `Value` and use the new method, you can also just copy this over from the `matrix-sdk-crypto-ffi` crate:

https://github.com/matrix-org/matrix-rust-sdk/blob/a11633381bf9232178c4522b1ff51ca03d081fc3/crates/matrix-sdk-crypto-ffi/src/machine.rs#L522

Reply:
Done!